### PR TITLE
custom instructions for copilot reviews

### PR DESCRIPTION
**Pull request by angiejones**
### Type of Change

- [ ] Feature
- [ ] Bug fix
- [ ] Refactor / Code quality
- [ ] Performance improvement
- [X] Documentation
- [ ] Tests
- [ ] Security fix
- [ ] Build / Release
- [ ] Other (specify below)

### AI Assistance

- [X] This PR was created or reviewed with AI assistance

### CODE REVIEW

**Review by michaelneale — APPROVED** [review]
looks good @angiejones where did you get those guidelines from? 

I think over time if we see it flag things incorrectly which are due to specific quirks of how goose works vs other style of apps, we can easily add the tips to here. 

I likeit.

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+- Off-by-one errors or boundary conditions
+- Incorrect error propagation (using `unwrap()` inappropriately)
+
+### Architecture & Patterns
+- Code that violates existing patterns in the codebase
+- Missing error handling (should use `anyhow::Result`)
+- Async/await misuse or blocking operations in async contexts
+- Improper trait implementations
+
+### Testing Gaps
```
Comment (DOsinga) [inline .github/copilot-instructions.md]: can we drop these? our current code base has way more tests with a negative value than tests that are missing; I would take bug fixes that don't have a matching test that make future regressions impossible

**Review by DOsinga — APPROVED** [review]
thanks for doing this, we do need something here. I wonder though if we should just all go through the reviews we got from copilot and collect what was useful and what was not.

**Comment by DOsinga** [comment]
fwiw, the three comments I got on this:

https://github.com/block/goose/pull/5643

were all spot on

**Comment by tlongwell-block** [comment]
This is a great idea. Something to keep in mind, copilot [does read](https://github.blog/changelog/2025-08-28-copilot-coding-agent-now-supports-agents-md-custom-instructions/) AGENTS.md in the root and in subdirectories. Though hard to tell if that is just the "coding" agent, or copilot in general, including the review bot.

(Maybe goose should check subdirs too when it is working. That would be a nice feature that would allow per-crate instructions, @DOsinga?)

So, we could probably reduce duplication a bit between the instructions file here and agents.

**Comment by DOsinga** [comment]
I actually have some notes on what goes often wrong in our code:

Do not make things optional that don't need to be. The compiler is annoying but is your friend
Remove useless LLM comments - at best they take up space, at worst they will be outdated soon and will confuse
Beware for Goose suggesting to check 3 different things. It is probably only one
Booleans should default to false and not be optional
Beware of the context manager that doesn't need context
Clean up your logging
Avoid overly defensive code

**Review by tlongwell-block — APPROVED** [review]

**Comment by tlongwell-block** [comment]
> I actually have some notes on what goes often wrong in our code:
> 
> Do not make things optional that don't need to be. The compiler is annoying but is your friend Remove useless LLM comments - at best they take up space, at worst they will be outdated soon and will confuse Beware for Goose suggesting to check 3 different things. It is probably only one Booleans should default to false and not be optional Beware of the context manager that doesn't need context Clean up your logging Avoid overly defensive code

I would probably put this in AGENTS.md so all AI working on the repo will follow these guidelines?

Looks like copilot review does respect agents, too https://docs.github.com/en/copilot/concepts/agents/code-review#providing-instructions-for-copilot-code-reviews

**Comment by angiejones** [comment]
> > I actually have some notes on what goes often wrong in our code:
> > Do not make things optional that don't need to be. The compiler is annoying but is your friend Remove useless LLM comments - at best they take up space, at worst they will be outdated soon and will confuse Beware for Goose suggesting to check 3 different things. It is probably only one Booleans should default to false and not be optional Beware of the context manager that doesn't need context Clean up your logging Avoid overly defensive code
> 
> I would probably put this in AGENTS.md so all AI working on the repo will follow these guidelines?
> 
> Looks like copilot review does respect agents, too https://docs.github.com/en/copilot/concepts/agents/code-review#providing-instructions-for-copilot-code-reviews

you want to move everything to AGENTS.md and delete the copilot custom instructions file, or just add these new entries to AGENTS.md?

**Comment by DOsinga** [comment]
it's a good point that we include AGENTS.md anyway, but I would still think repeating the important bits as explicit code review instructions (pay attention to optional things that shouldn't) wouldn't hurt

**Comment by angiejones** [comment]
> it's a good point that we include AGENTS.md anyway, but I would still think repeating the important bits as explicit code review instructions (pay attention to optional things that shouldn't) wouldn't hurt

added coding guidelines to AGENTS.md and review guidelines to the copilot custom instructions
https://github.com/block/goose/pull/5646/commits/2ccc3a795195364f0f7fc42ebcf2939de7f77fd0
